L1Trigger/TrackFindingTracklet/src/TrackletConfigBuilder.cc

```diff
@@ -53,18 +58,130 @@ TrackletConfigBuilder::TrackletConfigBuilder(const Settings& settings) : setting
+        // Convert layer number to Hybrid convention.
+        int layer = sm->layerId(); // Barrel = 1-6, Endcap = 11-15;
+        if (sm->barrel()) {
+          layer--; // Barrel 0-5
+        } else {
+          const int endcapOffsetHybrid = 5;
+          layer -= endcapOffsetHybrid; // Layer 6-19
+        }
+        // Inner radius of module.
+        float r = sm->r() - 0.5 * sm->numColumns()*sm->pitchCol()*fabs(sm->sinTilt());
+        // phi with respect to tracker nonant centre.
+        float phiMin = sm->phi() - 0.5*sm->numRows()*sm->pitchRow() / r;
+        float phiMax = sm->phi() + 0.5*sm->numRows()*sm->pitchRow() / r;
+        // Hybrid measures phi w.r.t. lower edge of tracker nonant.
+        const float phiOffsetHybrid = 0.5 * dphisectorHG_;
+        phiMin += phiOffsetHybrid;
+        phiMax += phiOffsetHybrid;
+        if (dtcPhiRange.find(layer) == dtcPhiRange.end()) {
+          dtcPhiRange[layer] = {phiMin, phiMax};
+        } else {
+          dtcPhiRange.at(layer).first  = std::min(phiMin, dtcPhiRange.at(layer).first);
+          dtcPhiRange.at(layer).second = std::max(phiMax, dtcPhiRange.at(layer).second);
+        }
+      }
+    }
+    for (const auto& p : dtcPhiRange) {
+      const unsigned int numSlots = setup->numATCASlots();
+      std::string dtcName = settings_.slotToDTCname(dtcId%numSlots);
+      if (dtcId >= numSlots) dtcName = "neg" + dtcName;
+      DTCinfo info;
+      info.name = dtcName;
+      info.layer = p.first;
+      info.phimin = p.second.first;
+      info.phimax = p.second.second;
+      vecDTCinfo_unsorted.push_back(info);
+    }
+  }
+
+  // Put DTCinfo vector in traditional order (PS first). (Needed?)
+  for (const DTCinfo& info : vecDTCinfo_unsorted) {
+    string dtcname = info.name;
+    if (dtcname.find("PS") != std::string::npos) {
+      vecDTCinfo_.push_back(info);
+    }
+  }
+  for (const DTCinfo& info : vecDTCinfo_unsorted) {
+    string dtcname = info.name;
+    if (dtcname.find("PS") == std::string::npos) {
+      vecDTCinfo_.push_back(info);
+    }
+  }
+}
+
+//--- Write DTC phi ranges to file to support stand-alone emulation.
+//--- (Only needed to support stand-alone emulation)
+
+void TrackletConfigBuilder::writeDTCphirange() const {
+  bool first = true;
+  for (const DTCinfo& info : vecDTCinfo_) {
+    string dirName = settings_.tablePath();
+    string fileName = dirName + "../dtcphirange.dat";
+    std::ofstream out;
+    openfile(out, first, dirName, fileName, __FILE__, __LINE__);
+    if (first) {
+      out<<"// phi ranges of modules read by each DTC"<<endl;
+      out<<"// (Used by stand-alone emulation)"<<endl;
+    }
+    out<<"vecDTCinfo_.push_back( {"""<<info.name<<""", "<<info.layer<<", "<<info.phimin<<", "<<info.phimax<<"} );"<<endl;
+    out.close();
+    first = false;
+  }
+}
+
+#else
+
+void TrackletConfigBuilder::setDTCphirange(const tt::Setup* setup) {
+#include "L1Trigger/TrackFindingTracklet/data/dtcphirange.txt"
```

Comment:
When running stand-alone, TrackletConfigBuilder must get the DTC phi ranges from hard-wired numbers, since it can't access the DB.  We must be able to update these hard-wired numbers easily from the DB whenever we want, so they don't get out of date. TrackletConfigBuilder::writeDTCphirange() provides this functionality.

I don't see another pretty solution if we want to keep stand-alone running. If we instead chose to leave these numbers hard-wired inside TrackletConfigBuilder.cc, it would be very difficult for writeDTCphirange() to update them. And we'd then have ugly constants in our code just to support the stand-alone running.

Reply:
I agree we should use a file - but I find writing files with code (called .txt) that are #included rather ugly. I would have written a file with just the values and then read in the values at runtime.